File: orange-pi-daemon/src/api_client.zig
const std = @import("std");
const json = std.json;
const types = @import("types.zig");

/// Sync mode for state reconciliation
pub const SyncMode = enum {
    full,
    incremental,
    
    pub fn toString(self: SyncMode) []const u8 {
        return switch (self) {
            .full => "full",
            .incremental => "incremental",
        };
    }
};

/// HTTP API client for uploading data to the dashboard
pub const ApiClient = struct {
    allocator: std.mem.Allocator,
    config: types.Config,
    client: std.http.Client,
    session_id: []const u8,

    pub fn init(allocator: std.mem.Allocator, config: types.Config) ApiClient {
        // Generate unique session ID for this daemon instance
        var prng = std.Random.DefaultPrng.init(@intCast(std.time.timestamp()));
        const random = prng.random();
        const session_id = std.fmt.allocPrint(allocator, "daemon-{x}-{x}", .{
            random.int(u64),
            @as(u64, @intCast(std.time.timestamp())),
        }) catch "daemon-unknown";
        
        std.log.info("🔑 Daemon session ID: {s}", .{session_id});
        
        return .{ 
            .allocator = allocator, 
            .config = config,
            .client = std.http.Client{ .allocator = allocator },
            .session_id = session_id,
        };
    }
    
    pub fn deinit(self: *ApiClient) void {
        self.allocator.free(self.session_id);
        self.client.deinit();
    }

    pub fn uploadPhone(self: *ApiClient, phone: types.Phone) !void {
        const phones = [_]types.Phone{phone};
        const phones_json = try std.json.Stringify.valueAlloc(self.allocator, phones, .{ .emit_null_optional_fields = false });
        defer self.allocator.free(phones_json);

        const payload = try std.fmt.allocPrint(self.allocator, "{{\"phones\":{s}}}", .{phones_json});
        defer self.allocator.free(payload);

        try self.makeRequest("/phones", payload);
        std.log.debug("✅ Uploaded phone {s} via HTTP API", .{phone.iccid});
    }

    pub fn uploadPhones(self: *ApiClient, phones: []const types.Phone) !void {
        if (phones.len == 0) return;
        

        const phones_json = try std.json.Stringify.valueAlloc(self.allocator, phones, .{ .emit_null_optional_fields = false });
        defer self.allocator.free(phones_json);

        const payload = try std.fmt.allocPrint(self.allocator, "{{\"phones\":{s}}}", .{phones_json});
        defer self.allocator.free(payload);

        try self.makeRequest("/phones", payload);
        std.log.debug("✅ Uploaded {d} phones via HTTP API", .{phones.len});
    }

    /// Upload device data with sync mode for state reconciliation
    pub fn uploadDevicesWithSync(self: *ApiClient, modems: []const types.Modem, sims: []const types.SIM, sync_mode: SyncMode) !void {
        if (modems.len == 0 and sims.len == 0) return;
        
        // Use the new JSON API (std.json.Stringify.valueAlloc)
        const modems_json = try std.json.Stringify.valueAlloc(self.allocator, modems, .{ .emit_null_optional_fields = false });
        defer self.allocator.free(modems_json);
        
        const sims_json = try std.json.Stringify.valueAlloc(self.allocator, sims, .{ .emit_null_optional_fields = false });
        defer self.allocator.free(sims_json);
        
        const timestamp = try std.fmt.allocPrint(self.allocator, "{}", .{std.time.timestamp()});
        defer self.allocator.free(timestamp);

        const payload = try std.fmt.allocPrint(self.allocator, 
            "{{\"sync_mode\":\"{s}\",\"session_id\":\"{s}\",\"timestamp\":\"{s}\",\"modems\":{s},\"sims\":{s}}}", 
            .{ sync_mode.toString(), self.session_id, timestamp, modems_json, sims_json }
        );
        defer self.allocator.free(payload);

        // Upload to the correct API endpoint
        try self.makeRequest("/devices", payload);
        
        std.log.info("✅ Uploaded {d} modems and {d} SIMs via HTTP API (mode: {s})", .{ modems.len, sims.len, sync_mode.toString() });
    }
    
    /// Legacy upload without sync mode (for backward compatibility)
    pub fn uploadDevices(self: *ApiClient, modems: []const types.Modem, sims: []const types.SIM) !void {
        try self.uploadDevicesWithSync(modems, sims, .incremental);
    }

    pub fn uploadMessages(self: *ApiClient, messages: []const types.Message) !void {
        if (messages.len == 0) return;
        

        std.log.info("📤 Uploading {d} new messages to API", .{messages.len});
        for (messages, 0..) |msg, i| {
            
            // Check if content is valid UTF-8
            const is_valid_utf8 = std.unicode.utf8ValidateSlice(msg.content);
            if (!is_valid_utf8) {
                std.log.warn("⚠️  Message {d} contains invalid UTF-8 and will be encoded as byte array", .{i});
                // Find the first invalid byte for debugging
                for (msg.content, 0..) |byte, pos| {
                    if (byte == 0xFE or byte == 0xFF) {
                        std.log.warn("⚠️  Found invalid byte 0x{X:0>2} at position {d}", .{ byte, pos });
                        break;
                    }
                }
            }
        }

        // Use the new JSON API (std.json.Stringify.valueAlloc)  
        const messages_json = try std.json.Stringify.valueAlloc(self.allocator, messages, .{ .emit_null_optional_fields = false });
        defer self.allocator.free(messages_json);

        const payload = try std.fmt.allocPrint(self.allocator, "{{\"messages\":{s}}}", .{messages_json});
        defer self.allocator.free(payload);

        try self.makeRequest("/messages", payload);
        std.log.info("✅ Uploaded {d} messages via HTTP API", .{messages.len});
    }

    pub fn getPendingSms(self: *ApiClient) ![]const types.PendingSms {

        const url = try std.fmt.allocPrint(self.allocator, "{s}/api/control/pending-sms", .{self.config.api_url});
        defer self.allocator.free(url);

        std.log.debug("🌐 Making GET request to: {s}", .{url});

        const uri = try std.Uri.parse(url);
        
        // Make the HTTP request using the correct Zig 0.15.1 API
        var request = try self.client.request(.GET, uri, .{
            .extra_headers = &[_]std.http.Header{
                .{ .name = "X-API-Key", .value = self.config.api_key },
                .{ .name = "Accept", .value = "application/json" },
            },
        });
        defer request.deinit();

        // Read response body using buffer approach  
        var response_body: std.ArrayList(u8) = .empty;
        defer response_body.deinit(self.allocator);
        
        // Read data in chunks
        var buffer: [4096]u8 = undefined;
        while (true) {
            const bytes_read = request.reader.interface.readSliceShort(&buffer) catch |err| switch (err) {
                error.ReadFailed => break,
                else => return err,
            };
            if (bytes_read == 0) break;
            try response_body.appendSlice(self.allocator, buffer[0..bytes_read]);
        }

        std.log.debug("📥 Response body: {s}", .{response_body.items});

        // Parse JSON response
        const parsed = try std.json.parseFromSlice(struct {
            success: bool,
            pending_messages: []types.PendingSms,
        }, self.allocator, response_body.items, .{ .ignore_unknown_fields = true });
        defer parsed.deinit();

        // Allocate and copy the pending SMS list
        const result_sms = try self.allocator.alloc(types.PendingSms, parsed.value.pending_messages.len);
        for (parsed.value.pending_messages, 0..) |sms, i| {
            result_sms[i] = .{
                .id = try self.allocator.dupe(u8, sms.id),
                .phone_iccid = try self.allocator.dupe(u8, sms.phone_iccid),
                .phone_number = if (sms.phone_number) |pn| try self.allocator.dupe(u8, pn) else null,
                .content = try self.allocator.dupe(u8, sms.content),
                .recipient = try self.allocator.dupe(u8, sms.recipient),
                .created_at = try self.allocator.dupe(u8, sms.created_at),
            };
        }

        return result_sms;
    }

    pub fn markSmsAsSent(self: *ApiClient, sms_id: []const u8) !void {

        // Use the correct endpoint and payload format
        const endpoint = "/sms-result";
        const payload = try std.fmt.allocPrint(self.allocator, 
            "{{\"message_id\":\"{s}\",\"success\":true}}", 
            .{sms_id}
        );
        defer self.allocator.free(payload);
        
        try self.makeRequest(endpoint, payload);
        std.log.info("✅ Marked SMS {s} as sent", .{sms_id});
    }

    fn makeRequest(self: *ApiClient, endpoint: []const u8, payload: []const u8) !void {
        const url = try std.fmt.allocPrint(self.allocator, "{s}/api/control{s}", .{ self.config.api_url, endpoint });
        defer self.allocator.free(url);

        std.log.debug("🌐 Making HTTP POST request to: {s}", .{url});
        std.log.debug("📤 Request payload: {s}", .{payload});

        // Use the fetch method for simpler implementation
        const result = self.client.fetch(.{
            .location = .{ .url = url },
            .method = .POST,
            .payload = payload,
            .extra_headers = &[_]std.http.Header{
                .{ .name = "Content-Type", .value = "application/json" },
                .{ .name = "X-API-Key", .value = self.config.api_key },
            },
        }) catch |err| {
            std.log.err("HTTP POST request failed: {any}", .{err});
            return error.HttpRequestFailed;
        };

        const status = @intFromEnum(result.status);
        std.log.debug("📥 HTTP Response Code: {d}", .{status});

        if (status < 200 or status >= 300) {
            std.log.err("HTTP request failed with status {d}", .{status});
            return error.HttpRequestFailed;
        }

        std.log.debug("📥 HTTP Response: Success", .{});
    }
};

File: orange-pi-daemon/src/api_client.rs
use crate::health::HealthSnapshot;
use crate::sync_manager::SyncMode;
use crate::types::*;
use anyhow::{Context, Result};
use reqwest::Client;
use serde_json::json;
use tracing::info;

#[derive(Clone)]
pub struct ApiClient {
    client: Client,
    pub config: Config,
}

impl ApiClient {
    pub fn new(config: Config) -> Self {
        Self {
            client: Client::builder()
                .timeout(std::time::Duration::from_secs(30)) // Increased to 30s for large uploads
                .build()
                .expect("Failed to create HTTP client"),
            config,
        }
    }

    /// Send an independent, versioned process and task health report.
    pub async fn send_health_snapshot(&self, snapshot: &HealthSnapshot) -> Result<()> {
        let url = format!("{}/api/control/heartbeat", self.config.api_url);
        let response = self
            .client
            .post(&url)
            .header("x-api-key", &self.config.api_key)
            .header("x-daemon-version", env!("CARGO_PKG_VERSION"))
            .json(snapshot)
            .send()
            .await
            .context("Failed to send health snapshot")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("Health endpoint returned {}: {}", status, body);
        }

        Ok(())
    }

    /// Upload changed modem reports and explicit removals.
    pub async fn upload_modem_reports(
        &self,
        reports: &[ModemReport],
        removed_equipment_ids: &[String],
        sync_mode: SyncMode,
        session_id: &str,
    ) -> Result<()> {
        if sync_mode == SyncMode::Incremental
            && reports.is_empty()
            && removed_equipment_ids.is_empty()
        {
            return Ok(());
        }

        let url = format!("{}/api/control/devices", self.config.api_url);
        let timestamp = chrono::Utc::now()
            .format("%Y-%m-%dT%H:%M:%S%.3fZ")
            .to_string();

        info!(
            "📤 Uploading {} changed modem reports and {} removals (mode: {}, session: {})",
            reports.len(),
            removed_equipment_ids.len(),
            sync_mode.as_str(),
            session_id
        );

        let payload = json!({
            "sync_mode": sync_mode.as_str(),
            "session_id": session_id,
            "timestamp": timestamp,
            "modem_reports": reports,
            "removed_equipment_ids": removed_equipment_ids,
        });

        let response = self
            .client
            .post(&url)
            .header("x-api-key", &self.config.api_key)
            .header("x-daemon-version", "rust-3.0.0-reports")
            .json(&payload)
            .send()
            .await
            .context("Failed to send modem reports")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response
                .text()
                .await
                .unwrap_or_else(|_| String::from("(no body)"));
            anyhow::bail!("API returned error: {} - {}", status, body);
        }

        info!(
            "✅ Successfully uploaded modem reports (mode: {})",
            sync_mode.as_str()
        );
        Ok(())
    }

    /// Upload messages with proper batching to avoid overwhelming the API
    pub async fn upload_messages(&self, messages: &[Message]) -> Result<()> {
        if messages.is_empty() {
            return Ok(());
        }

        const BATCH_SIZE: usize = 50; // Match the server's internal batch size
        let url = format!("{}/api/control/messages", self.config.api_url);

        // Process messages in batches to avoid overwhelming the API
        let total = messages.len();
        let mut uploaded = 0;

        for (batch_num, chunk) in messages.chunks(BATCH_SIZE).enumerate() {
            info!(
                "📤 Uploading batch {}/{} ({} messages)",
                batch_num + 1,
                (total + BATCH_SIZE - 1) / BATCH_SIZE,
                chunk.len()
            );

            // Upload this batch
            let response = self
                .client
                .post(&url)
                .header("x-api-key", &self.config.api_key)
                .json(&json!({ "messages": chunk }))
                .send()
                .await
                .context(format!("Failed to upload message batch {}", batch_num + 1))?;

            if !response.status().is_success() {
                let status = response.status();
                let body = response
                    .text()
                    .await
                    .unwrap_or_else(|_| String::from("(no body)"));
                anyhow::bail!(
                    "API returned error for batch {}: {} - {}",
                    batch_num + 1,
                    status,
                    body
                );
            }

            uploaded += chunk.len();

            // Small delay between batches to avoid rate limiting
            if batch_num < messages.chunks(BATCH_SIZE).len() - 1 {
                tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
            }
        }

        info!(
            "✅ Successfully uploaded {} messages in {} batches",
            uploaded,
            (total + BATCH_SIZE - 1) / BATCH_SIZE
        );
        Ok(())
    }

    /// Get pending SMS to send
    pub async fn get_pending_sms(&self) -> Result<Vec<PendingSms>> {
        let url = format!("{}/api/control/pending-sms", self.config.api_url);

        let response = self
            .client
            .get(&url)
            .header("x-api-key", &self.config.api_key)
            .send()
            .await
            .context("Failed to get pending SMS")?;

        if !response.status().is_success() {
            anyhow::bail!("API returned error: {}", response.status());
        }

        #[derive(serde::Deserialize)]
        struct PendingResponse {
            pending_messages: Vec<PendingSms>,
        }

        let data: PendingResponse = response.json().await?;
        Ok(data.pending_messages)
    }
}
